### PR TITLE
media-gfx/opencsg: bump to 1.5.1

### PR DESCRIPTION
**Pull request by waebbl**
Closes: https://bugs.gentoo.org/838439
Signed-off-by: Bernd Waibel <waebbl-gentoo@posteo.net>

### CODE REVIEW

File: media-gfx/opencsg/opencsg-1.5.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+RDEPEND="
+	dev-libs/libbsd
+	media-libs/glew:0=
+	virtual/opengl
+	x11-libs/libX11
+	x11-libs/libXau
+	x11-libs/libXdmcp
+	x11-libs/libXext
+	x11-libs/libxcb:=
+"
```
Comment (juippis) [inline media-gfx/opencsg/opencsg-1.5.0.ebuild:30]:
```
 * VDB: detected possibly incorrect RDEPEND (media-gfx/opencsg-1.5.0)
 * dev-libs/libbsd   < 
 * x11-libs/libX11   < 
 * x11-libs/libXau   < 
 * x11-libs/libXdmcp < 
 * x11-libs/libXext  < 
 * x11-libs/libxcb:= < 
```
Could you check? It means these libraries aren't directly linked to 
```
# scanelf -n /usr/lib64/libopencsg.so.1.5.0
 TYPE   NEEDED FILE 
ET_DYN libGLEW.so.2.2,libGL.so.1,libstdc++.so.6,libgcc_s.so.1,libc.so.6 /usr/lib64/libopencsg.so.1.5.0
```

Comment (waebbl) [reply media-gfx/opencsg/opencsg-1.5.0.ebuild:30]: I'm gonna verify those

Comment (waebbl) [reply media-gfx/opencsg/opencsg-1.5.0.ebuild]: They were indeed not needed to build the package.

**Comment by waebbl** [comment]
@juippis Please don't merge yet. Upstream just released 1.5.1 today, which already contains the fix from the patch. I'd like to update the package to 1.5.1 first to avoid the added patch.

**Review by juippis — COMMENTED** [review]
Oh no...
```
# qa-cmp -I media-gfx/opencsg
CMP: media-gfx/opencsg-1.4.2-r1/image with media-gfx/opencsg-1.5.1/image
 FILES:-usr/lib64/libopencsg.so.1.4
 FILES:+usr/lib64/libopencsg.so.1.5
   ABI: libopencsg.so.1(64) func(-1) [BREAKING]
------> FILES(+1,-1) ABI(-1,>B<)
```
Good news is though:
```
$ rdep media-gfx/opencsg
== rdep of media-gfx/opencsg ==
media-gfx/openscad-2021.01
media-gfx/openscad-9999
== ddep of media-gfx/opencsg ==
media-gfx/openscad-2021.01
media-gfx/openscad-9999
== bdep of media-gfx/opencsg ==
== pdep of media-gfx/opencsg ==
```
We'll have to add subslot for this package and revbump the media-gfx/openscad-2021.01 rdep adding `media-gfx/opencsg:=`.

**Comment by waebbl** [comment]
Ok, I wanted to look at the openscad-2021.01 anyway, because it still has an open bug AFAIR.

Is it ok to add these changes to this PR? They are related in some way.
Is the ABI breakage sth that should be reported upstream? AFAIK ABI breaks shouldn't happen between minor releases, or do I remember this wrong?

**Comment by juippis** [comment]
> Ok, I wanted to look at the openscad-2021.01 anyway, because it still has an open bug AFAIR.
> 

Remember that the only version in-tree is stable.


> Is it ok to add these changes to this PR? They are related in some way. 

Yes please, that is preferred.


> Is the ABI breakage sth that should be reported upstream? AFAIK ABI breaks shouldn't happen between minor releases, or do I remember this wrong?

This is 1.4.2 -> 1.5.1 jump, because 1.5.0 wasn't packaged. But I'm pretty sure 1.4.2 -> 1.5.0 breaks ABI too. Sure you can ask upstream about it, but it seems pretty normal to me.

**Review by juippis — APPROVED** [review]
Thanks! We should add the subslot binder to the stable ebuild too so it won't hit us accidentally later on. I'll do some git mv'ing.

**Comment by waebbl** [comment]
Thanks for the git moving. I would have aimed for a quick stabilization to reduce potential issues with the stable ebuild.

File: media-gfx/opencsg/opencsg-1.5.1.ebuild [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+MY_P="OpenCSG-${PV}"
+
+DESCRIPTION="The Constructive Solid Geometry rendering library"
+HOMEPAGE="https://www.opencsg.org"
+SRC_URI="https://www.opencsg.org/${MY_P}.tar.gz"
+S="${WORKDIR}/${MY_P}"
+
+LICENSE="GPL-2+"
+SLOT="0"
```
Comment (thesamesam) [inline media-gfx/opencsg/opencsg-1.5.1.ebuild:16]: This is still 0...?

Comment (waebbl) [reply media-gfx/opencsg/opencsg-1.5.1.ebuild:16]: Is there a reason why it should change?

Comment (thesamesam) [reply media-gfx/opencsg/opencsg-1.5.1.ebuild:16]: ABI changed, so the subslot needs to represent that to trigger rebuilds? Or am I misunderstanding what happened here(

Comment (waebbl) [reply media-gfx/opencsg/opencsg-1.5.1.ebuild:16]: Isn't this why I added the subslot operator? Should the update in version number together with this operator enough to trigger rebuilds?

Comment (thesamesam) [reply media-gfx/opencsg/opencsg-1.5.1.ebuild:16]: Subslot operators do nothing if the subslot didn't change. They don't work based on version.

Comment (juippis) [reply media-gfx/opencsg/opencsg-1.5.1.ebuild:16]: Eeh woops, indeed. I'll sort the tree!

Comment (thesamesam) [reply media-gfx/opencsg/opencsg-1.5.1.ebuild:16]: Thanks!

Comment (juippis) [reply media-gfx/opencsg/opencsg-1.5.1.ebuild:16]: So yeah @waebbl `SLOT="0/1.5" 
0 = slot operator,
1.5 = subslot

when subslot changes it means the new SONAME is incompatible and every ebuild depending on your `package:=` will get rebuilt. Think it this way: openscad is now **bound** to opencsg:0/1.5. 

Whenever you update opencsg you should check if the SONAME has changed, and/or if the new libs break the API/ABI. 
```
# objdump -p /usr/lib64/libopencsg.so.1.5.1 | grep SONAME
  SONAME               libopencsg.so.1
```
note that it's still called "1" but I named it "1.5" in anticipation of it breaking with 1.6 release. If 1.6.0 is fine, you can **keep** the "0/1.5" slot. That is perfectly fine. 

Here's how you easiest check the SONAME API/ABI breakage:

> emerge -1av iwdevtools libabigail
> (if you're not, *cd* to your ebuild development directory ../media-gfx/opencsg)
> ebuild opencsg-oldversion clean install
> ebuild opencsg-newversion clean install
> qa-cmp -I media-gfx/opencsg
```
 CMP: media-gfx/opencsg-1.4.2-r1/image with media-gfx/opencsg-1.5.1/image
 FILES:-usr/lib64/libopencsg.so.1.4
 FILES:+usr/lib64/libopencsg.so.1.5
   ABI: libopencsg.so.1(64) func(-1) [BREAKING]
------> FILES(+1,-1) ABI(-1,>B<)
```

https://devmanual.gentoo.org/general-concepts/slotting/index.html#sub-slots

Comment (waebbl) [reply media-gfx/opencsg/opencsg-1.5.1.ebuild:16]: Thanks @thesamesam and @juippis for fixing this on the fly and detailed explanation.

**Pull request closed without merging**